src/Projection/AbstractProjection.php

```diff
@@ -133,10 +131,13 @@ protected function handleStreamWithHandlers(string $streamName, Iterator $events
         foreach ($events as $event) {
             /* @var Message $event */
             $this->position->inc($streamName);
+
             if (! isset($this->handlers[$event->messageName()])) {
                 continue;
             }
+
             $handler = $this->handlers[$event->messageName()];
+            $handler = Closure::bind($handler, $this->createHandlerContext($streamName));
```

Comment:
handler context could be created outside of the `foreach` and with a `setStreamName` method but I would only do this if it has really positive influence on the performance.